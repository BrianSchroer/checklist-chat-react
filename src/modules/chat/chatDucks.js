// Chat actions/reducer "ducks" file (https://github.com/erikras/ducks-modular-redux)

import initialState from '../../app/store/initialState';
import {beginAjaxCall, ajaxCallError} from '../../app/ajaxStatusDucks';
import * as mockSignalR from '../../api/mockSignalR';

const prefix = 'checklist-chat/chat/';

// won't be used with SignalR?:
export const LOAD_CHAT_MESSAGES_FOR_ROOM_SUCCESS = `${prefix}LOAD_CHAT_MESSAGES_FOR_ROOM_SUCCESS`;

export const DISPLAY_CHAT_MESSAGE = `${prefix}DISPLAY_CHAT_MESSAGE`;
export const DISPLAY_CHAT_USER_ACTION = `${prefix}DISPLAY_CHAT_USER_ACTION`;
export const DISPLAY_PRIORITY_NOTIFICATION = `${prefix}DISPLAY_PRIORITY_NOTIFICATION`;
export const UPDATE_CHAT_PARTICIPANTS = `${prefix}UPDATE_CHAT_PARTICIPANTS`;

// Actions:

export const loadChatMessagesForRoomSuccess =
    (chatMessages) => ({type: LOAD_CHAT_MESSAGES_FOR_ROOM_SUCCESS, chatMessages});

export const updateChatParticipants =
    (chatParticipants) => ({type: UPDATE_CHAT_PARTICIPANTS, chatParticipants});

const updateMessagesAndParticipants = (chatMessages, dispatch) => {
    dispatch(loadChatMessagesForRoomSuccess(chatMessages));

    const sortedNames = chatMessages.map(m => m.userName).sort();
    const uniqueNames = [...new Set(sortedNames)];

    const chatParticipants = uniqueNames.map(userName => ({
        id: userName,
        name: userName,
        department: 'Department',
        title: 'Title',
        connection: 'Connection'
    }));

    dispatch(updateChatParticipants(chatParticipants));
};

export function loadChatMessagesForRoom(roomId) {
    if (roomId) {
        return dispatch => {
            dispatch(beginAjaxCall());

            return mockSignalR.getChatMessages(roomId).then(chatMessages =>
                updateMessagesAndParticipants(chatMessages, dispatch)
            ).catch(error => {
                dispatch(ajaxCallError(error));
                throw (error);
            });
        };
    } else {
        return [];
    }
}

export function saveChatMessage(chatMessage, roomId) {
    return dispatch => {
        dispatch(beginAjaxCall());

        return mockSignalR.chat(roomId, chatMessage).then(() =>
        {
            mockSignalR.getChatMessages(roomId).then(chatMessages =>
                updateMessagesAndParticipants(chatMessages, dispatch));
        }).catch(error => {
            dispatch(ajaxCallError(error));
            throw (error);
        });
    };
}

// Reducers:

export function chatMessagesReducer(chatMessages = initialState.chatMessages, action) {
    const actionType = action.type;

    switch (actionType) {

        case LOAD_CHAT_MESSAGES_FOR_ROOM_SUCCESS:
            return action.chatMessages;

        case DISPLAY_CHAT_MESSAGE:
        case DISPLAY_CHAT_USER_ACTION:
            return [...chatMessages, Object.assign({}, action.chatMessage)];

        default:
            return chatMessages;
    }
}

export function chatParticipantsReducer(chatParticipants = initialState.chatParticipants, action) {
    const actionType = action.type;

    switch (actionType) {

        case UPDATE_CHAT_PARTICIPANTS:
            return action.chatParticipants;

        default:
            return chatParticipants;
    }
}
