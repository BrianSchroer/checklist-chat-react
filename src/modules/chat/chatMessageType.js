export const CHAT = 'Chat';
export const ACTION = 'Action';
