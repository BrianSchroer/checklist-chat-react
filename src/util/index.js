import format from './format';
import uiHelper from './uiHelper';
import validationHelper from './validationHelper';

export {
    format as format,
    uiHelper as uiHelper,
    validationHelper as validationHelper
};
