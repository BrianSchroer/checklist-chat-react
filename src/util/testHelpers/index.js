import {shallow, mount} from 'enzyme';
import enzymeHelper from './enzymeHelper';

export {
    shallow as shallow,
    mount as mount,
    enzymeHelper as enzymeHelper
};
