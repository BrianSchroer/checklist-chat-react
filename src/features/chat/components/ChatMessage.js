import React, {PropTypes} from 'react';
import format from '../../../util/format';
import * as chatMessageType from '../chatMessageType';

const ChatMessage = ({chatMessage}) => {
    switch (chatMessage.chatMessageType) {
        case chatMessageType.ACTION:
            return (
                <div className="chat-action-message">
                    <div className="timestamp chat-message-timestamp">
                        {format.time(chatMessage.timeStamp)}
                    </div>
                    <span className="chat-action-message-text">{`${chatMessage.userName} ${chatMessage.text}`}</span>
                </div>
            );
        case chatMessageType.CHAT:
        default:
            return (
                <div className="chat-message">
                    <div className="timestamp chat-message-timestamp">
                        {format.time(chatMessage.timeStamp)}
                    </div>
                    <strong>{`${chatMessage.userName}: `}</strong>
                    <span className="chat-message-text">{chatMessage.text}</span>
                </div>
            );
    }
};

ChatMessage.propTypes = {
    chatMessage: PropTypes.object.isRequired
};

export default ChatMessage;
