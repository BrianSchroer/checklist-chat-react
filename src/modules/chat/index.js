import * as chatMessageType from './chatMessageType';
import ChatMessageMarkdown from './components/ChatMessageMarkdown';

export {
    chatMessageType as chatMessageType,
    ChatMessageMarkdown as ChatMessageMarkdown
};
