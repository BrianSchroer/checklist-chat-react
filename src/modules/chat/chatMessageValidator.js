import {validationHelper} from '../../util';

export function validate(chatMessage) {
    return validationHelper.validateFields([{
        fieldName: 'text',
        predicate: validationHelper.missingRequiredValue(chatMessage.text),
        errorMessage: 'Message is required.'
    }]);
}
