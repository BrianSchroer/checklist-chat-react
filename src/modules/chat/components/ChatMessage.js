import React from 'react';
import PropTypes from 'prop-types';
import {format} from '../../../util';
import {chatMessageType} from '../../chat';

const withPriorityClass = (className, chatMessage, userId) => {
    let priorityClass = className;

    if (userId) {
        const {priorityNotificationRecipients} = chatMessage;
        if (priorityNotificationRecipients && priorityNotificationRecipients.includes(userId)) {
            priorityClass += ' high-priority';
        }
    }

    return priorityClass;
};

const ChatMessage = ({chatMessage, userId}) => {
    switch (chatMessage.chatMessageType) {
        case chatMessageType.ACTION:
            return (
                <div className="chat-action-message text-muted">
                    <div className="timestamp chat-message-timestamp">
                        {format.time(chatMessage.timeStamp)}
                    </div>
                    <span className="chat-action-message-text">{`${chatMessage.userName} ${chatMessage.text}`}</span>
                </div>
            );

        case chatMessageType.CHAT:
        default:
            return (
                <div className={withPriorityClass('chat-message', chatMessage, userId)}>
                    <div className="timestamp chat-message-timestamp">
                        {format.time(chatMessage.timeStamp)}
                    </div>
                    <strong>{`${chatMessage.userName}: `}</strong>
                    <span className="chat-message-text">{chatMessage.text}</span>
                </div>
            );
    }
};

ChatMessage.propTypes = {
    chatMessage: PropTypes.object.isRequired,
    userId: PropTypes.string
};

export default ChatMessage;
