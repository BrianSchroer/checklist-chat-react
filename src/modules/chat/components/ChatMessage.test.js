import React from 'react';
import {snapshotHelper} from '../../../util/testHelpers';
import ChatMessage from './ChatMessage';
import {chatMessageType} from '../../chat';

const testMessage = {
    timeStamp: '2016-12-08T14:57:10.222Z',
    userName: 'test userName',
    text: 'test text'
};

function overrideMessage(messageOverrides) {
    return {
        userId: 'currentUser',
        chatMessage: Object.assign({}, testMessage, messageOverrides)
    };
}

function assertSnapshotMatch(messageOverrides) {
    snapshotHelper.assertMatch(<ChatMessage {...overrideMessage(messageOverrides)}/>);
}

describe('ChatMessage', () => {
    describe(`when chatMessageType = "${chatMessageType.ACTION}"`, () => {
        it('should render correctly', () => {
            assertSnapshotMatch({chatMessageType: chatMessageType.ACTION});
        });
    });

    describe(`when chatMessageType = "${chatMessageType.CHAT}"`, () => {
        it('should render correctly', () => {
            assertSnapshotMatch({chatMessageType: chatMessageType.CHAT});
        });

        it('should render priority notification for current recipient with "high-priority" class', () => {
            assertSnapshotMatch({
                chatMessageType: chatMessageType.CHAT,
                priorityNotificationRecipients: ['currentUser', 'anotherUser']
            });
        });

        it('should render priority notification for other recipients without "high-priority" class', () => {
            assertSnapshotMatch({
                chatMessageType: chatMessageType.CHAT,
                priorityNotificationRecipients: ['anotherUser', 'yetAnotherUser']
            });
        });
    });
});
