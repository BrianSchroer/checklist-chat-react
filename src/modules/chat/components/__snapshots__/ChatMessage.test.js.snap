// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatMessage when chatMessageType = "Action" should render correctly 1`] = `
<div
  className="chat-action-message text-muted"
>
  <div
    className="timestamp chat-message-timestamp"
  >
    8:57 am
  </div>
  <span
    className="chat-action-message-text"
  >
    test userName test text
  </span>
</div>
`;

exports[`ChatMessage when chatMessageType = "Chat" should render correctly 1`] = `
<div
  className="chat-message"
>
  <div
    className="timestamp chat-message-timestamp"
  >
    8:57 am
  </div>
  <strong>
    test userName: 
  </strong>
  <span
    className="chat-message-text"
  >
    test text
  </span>
</div>
`;

exports[`ChatMessage when chatMessageType = "Chat" should render priority notification for current recipient with "high-priority" class 1`] = `
<div
  className="chat-message high-priority"
>
  <div
    className="timestamp chat-message-timestamp"
  >
    8:57 am
  </div>
  <strong>
    test userName: 
  </strong>
  <span
    className="chat-message-text"
  >
    test text
  </span>
</div>
`;

exports[`ChatMessage when chatMessageType = "Chat" should render priority notification for other recipients without "high-priority" class 1`] = `
<div
  className="chat-message"
>
  <div
    className="timestamp chat-message-timestamp"
  >
    8:57 am
  </div>
  <strong>
    test userName: 
  </strong>
  <span
    className="chat-message-text"
  >
    test text
  </span>
</div>
`;
